pkg/serviceaccounts/controllers/create_dockercfg_secrets.go

```diff
@@ -298,62 +414,75 @@ func (e *DockercfgController) syncServiceAccount(key string) error {
 	return err
 }
 
-const (
-	tokenSecretWaitInterval = 20 * time.Millisecond
-	tokenSecretWaitTimes    = 100
-)
+// getTokenSecretFromCache retrieve the token secret from the cache
+func (e *DockercfgController) getTokenSecretFromCache(serviceAccount *api.ServiceAccount) (*api.Secret, bool, error) {
+	// If the token secret was already created for this service account but it is not yet
+	// available in the cache wait for the shared informer to sync.
+	// When the token already exists in the cache and it is populated indicate that to
+	// createDockercfgSecrets and proceed with dockercfg generation.
+	if e.pendingTokens.Has(serviceAccount) {
+		return nil, true, nil
+	}
+	secrets, err := e.secretsIndex.ByIndex("serviceAccounts", makeKeyFromServiceAccount(serviceAccount))
+	if err != nil {
+		return nil, false, err
+	}
+	for _, obj := range secrets {
+		secret := obj.(*api.Secret)
+		if createdBy := secret.Annotations[api.CreatedByAnnotation]; createdBy == CreateDockercfgSecretsController {
+			return secret, false, nil
+		}
+	}
+	return nil, false, nil
+}
 
 // createTokenSecret creates a token secret for a given service account.  Returns the name of the token
-func (e *DockercfgController) createTokenSecret(serviceAccount *api.ServiceAccount) (*api.Secret, error) {
-	tokenSecret := &api.Secret{
+func (e *DockercfgController) createTokenSecret(serviceAccount *api.ServiceAccount) (*api.Secret, bool, error) {
+	tokenSecret, isPending, err := e.getTokenSecretFromCache(serviceAccount)
```

Comment:
pretty sure looking up the token from cache here is a race:

1. service account exists, token exists, dockercfg exists, serviceaccount has reference, caches are populated
2. token gets deleted
3. deleted_token_secrets.go deletes the dockercfg in response
4. deleted_dockercfg_secrets.go removes the reference from the service account in response
5. service account update triggers this controller to resync the service account. this controller's token cache could still show the populated token that was deleted in step 2, and it would re-create a dockercfg secret now containing a defunct token and add it to the serviceaccount